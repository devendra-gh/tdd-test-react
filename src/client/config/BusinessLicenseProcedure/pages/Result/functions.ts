import { IVariables } from '@tamm/app-composer';
import InternalApi from 'client/services/InternalApi';
import moment from 'moment';
import fetchFees from '../../services/fees';
import fetchRequirements from '../../services/requirements';
import { initialState } from '../../config';
import { getAnalyticsData } from '../../utils';

const getInitialState = async (props: IVariables) => {
  const {
    actions,
    resultState,
    formCompanyDetails,
    formSelectActivity,
  } = props;

  const { location, legalForm } = formCompanyDetails;
  const { activity } = formSelectActivity;
  const payload = { location, legalForm, activityId: activity };
  actions.resultState.update(initialState.resultState);
  Promise.all([fetchFees(payload), fetchRequirements(payload)])
    .then(([fees, requirements]) => {
      actions.resultState.update({
        ...resultState,
        loading: false,
        showError: false,
        fees,
        requirements,
      });
      getAnalyticsData('sla', { serviceStatus: 'success' });
    })
    .catch(err => {
      props.history.push('/business-licence-procedure/error-page');
      actions.resultState.update({
        ...resultState,
        loading: false,
        showError: true,
      });
      getAnalyticsData('sla', { serviceStatus: 'fail' });
    });
};
const getRequirements = (props: IVariables) => {
  const {
    locale,
    resultState: { requirements },
  } = props;
  return (
    (requirements &&
      requirements.length &&
      requirements.map((requirement: any) => {
        return {
          requirement:
            locale === 'en'
              ? requirement.requirementDescEn
              : requirement.requirementDescAr,
        };
      })) ||
    []
  );
};

const getFees = (props: IVariables) => {
  const {
    locale,
    resultState: { fees },
  } = props;
  return (
    (fees &&
      fees.length &&
      fees.map((fee: any) => {
        return {
          desc: locale === 'en' ? fee.feeDescEn : fee.feeDescAr,
          status: fee.activityCode,
          fee: fee.feeAmount,
        };
      })) ||
    []
  );
};

const getDownloadLicenseProcedureDoc = async (props: IVariables) => {
  const {
    i18n,
    locale,
    resultState: { legalForm, legalFormAr, selectedActivity, fees },
  } = props;

  const payload = {
    printDate: moment(new Date()).format('DD/MM/YYYY'),
    printTime: moment(new Date()).format('hh:mm A'),
    transactionType: i18n('applicant_new_license'),
    legalType: locale === 'en' ? legalForm : legalFormAr,
    activities: [
      {
        activity:
          locale === 'en'
            ? selectedActivity.activity
            : selectedActivity.activityAr,
      },
    ],
    conditions: [
      {
        condition: i18n('applicant_age_constant'),
      },
    ],
    requirements: getRequirements(props),
    fees: getFees(props),
    total:
      fees && fees.length
        ? fees.reduce(
            (accumulator: number, currentValue: IVariables) =>
              accumulator + currentValue.feeAmount,
            0,
          )
        : '',
  };
  const docxData = { payload, locale };

  try {
    const response = await InternalApi.getDownloadLicenseProcedureDoc(docxData);
    if (response && response.data) {
      const filename = 'Certificate';
      const typedArr = new Uint8Array(response.data.data);
      const stringChar = typedArr.reduce((data, byte) => {
        return data + String.fromCharCode(byte);
      }, '');
      const base64String = btoa(stringChar);
      const data = `data:application/pdf;base64,${base64String}`;
      const link = document.createElement('a');
      link.href = data;
      link.download = `${filename}.pdf`;
      link.click();
    }
  } catch (err) {
    // console.error('Error while getting document:', err);
  }
};

export default {
  getInitialState,
  getDownloadLicenseProcedureDoc,
};
