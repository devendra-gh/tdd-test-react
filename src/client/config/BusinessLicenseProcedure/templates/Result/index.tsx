import React, { useEffect } from 'react';
import { withTemplateHooks, IVariables } from '@tamm/app-composer';
import Container from 'client/containers';
import Sidebar from 'client/templates/Sidebar';
import Table from '@tamm/ui-lib-v2-table';
import Form from '@tamm/ui-lib-v2-form';
import Button from '@tamm/ui-lib-v2-button';
import Loading from '../../components/Loading';
import { noRefCheck } from '../../utils';

function Result(props: IVariables) {
  const {
    i18n,
    subTitle,
    description,
    locale,
    resultState,
    getDownloadLicenseProcedureDoc,
  } = props;

  const { requirements, fees, loading, selectedActivity } = resultState;

  useEffect(() => {
    props.getInitialState(props);
  }, []);

  return (
    <Container
      locale={props.locale}
      sidebar={
        <Sidebar
          currentStep={props.currentStep}
          currentSubStep={props.currentSubStep}
          i18n={i18n}
          stepsStatus={props.stepsStatus}
          steps={props.steps}
        />
      }
    >
      <div className="">
        <h3
          className="syb-title"
          style={{ marginBottom: '2rem', color: '#161038' }}
        >
          {i18n(subTitle)}
        </h3>
        <p style={{ color: '#3f3e45' }}>{i18n(description)}</p>
      </div>
      <div style={{ height: '40px' }} />
      {loading && <Loading />}
      {!loading && selectedActivity && (
        <>
          <Table
            i18n={i18n}
            columns={[
              {
                id: 'name',
                title: i18n('table.header.activityName'),
                render: () =>
                  locale === 'en'
                    ? selectedActivity.activity
                    : selectedActivity.activityAr || '',
              },
              {
                id: 'total',
                title: i18n('table.header.totalFees'),
                align: 'end',
                render: () =>
                  i18n('global.aed', {
                    amount:
                      fees && fees.length
                        ? fees.reduce(
                            (accumulator: number, currentValue: IVariables) =>
                              accumulator + currentValue.feeAmount,
                            0,
                          )
                        : '',
                  }),
              },
            ]}
            items={[
              {
                id: '1',
                label: '',
              },
            ]}
            onClick={noRefCheck}
            onSelectionChange={noRefCheck}
            size="default"
            title={i18n('table.heading.licenceSummary')}
          />
          <div style={{ height: '40px' }} />
          {fees && !!fees.length && (
            <Table
              i18n={i18n}
              columns={[
                {
                  id: 'description',
                  title: i18n('table.header.feeDescription'),
                },
                {
                  id: 'fee',
                  title: i18n('table.header.estimatedFee'),
                  align: 'end',
                },
              ]}
              items={
                fees &&
                fees.map((fee: IVariables) => ({
                  description: locale === 'en' ? fee.feeDescEn : fee.feeDescAr,
                  id: locale === 'en' ? fee.feeDescEn : fee.feeDescAr,
                  fee: i18n('global.aed', { amount: fee.feeAmount }),
                  selected: false,
                }))
              }
              onClick={noRefCheck}
              size="default"
              title={i18n('table.heading.feeSummary')}
            />
          )}
          <div style={{ height: '40px' }} />
          {requirements && !!requirements.length && (
            <Table
              i18n={i18n}
              columns={[
                {
                  id: 'description',
                  title: i18n('table.header.requirement'),
                },
                {
                  id: 'type',
                  title: i18n('table.header.requirementType'),
                },
                {
                  id: 'authority',
                  title: i18n('table.header.authority'),
                },
              ]}
              items={
                requirements &&
                requirements.map((requirement: IVariables) => ({
                  description:
                    locale === 'en'
                      ? requirement.requirementDescEn
                      : requirement.requirementDescAr,
                  id: requirement.requirementId,
                  type:
                    locale === 'en'
                      ? requirement.requirementTypeEn
                      : requirement.requirementTypeAr,
                  authority:
                    locale === 'en'
                      ? requirement.authorityEn
                      : requirement.authorityAr,
                  selected: false,
                }))
              }
              onClick={noRefCheck}
              size="default"
              title={i18n('table.heading.requirements')}
            />
          )}
          <div
            style={{
              display: 'flex',
              width: '120px',
              justifyContent: 'space-between',
              marginTop: '40px',
              marginBottom: '40px',
            }}
          >
            <Button
              aria-label="print"
              label={i18n('button.print')}
              onClick={() => window.print()}
              uiType="secondary"
            />
            <span style={{ marginRight: '1rem', marginLeft: '1rem' }} />
            <Button
              aria-label="download"
              label={i18n('button.download')}
              onClick={() => getDownloadLicenseProcedureDoc(props)}
              uiType="secondary"
            />
          </div>
        </>
      )}
      <div style={{ height: '40px' }} />

      <Form
        backButton={{
          label: props.i18n('button.back'),
          withArrow: true,
          alignIcon: 'start',
          uiType: 'secondary',
          onClick: () =>
            props.history.push('/business-licence-procedure/search-activities'),
        }}
      />
      <div style={{ height: '40px' }} />
    </Container>
  );
}

export default withTemplateHooks(Result);
// export default Result;
