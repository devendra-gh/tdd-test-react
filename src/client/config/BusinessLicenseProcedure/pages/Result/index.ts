import { IVariables } from '@tamm/app-composer';
import functions from './functions';

const routes = [
  {
    path: ['/business-licence-procedure/licence-information/'],
    uniqueId: 'businessLicenseProcedure-SelectCompanyDetails',
    template: 'businessLicenseProcedureResult',
    title: 'businessLicenseProcedure.main.title',
    props: {
      currentStep: 'businessLicenseProcedure.step3.title',
      getInitialState: functions.getInitialState,
      getDownloadLicenseProcedureDoc: functions.getDownloadLicenseProcedureDoc,
      onSubmitLabel: 'button.next',
      withArrow: true,
      subTitle: 'businessLicenseProcedure.step3.title',
      description: 'businessLicenseProcedure.step3.page.description',
      stepsStatus: {
        'businessLicenseProcedure.step1.title': 'finish',
        'businessLicenseProcedure.step2.title': 'finish',
      },
    },
    init: (props: IVariables) => {
      const { formCompanyDetails, formSelectActivity, actions } = props;
      const { location, legalForm } = formCompanyDetails;
      const { activities, activity } = formSelectActivity;
      if (!activities) {
        props.history.push('/business-licence-procedure/error-page');
        return;
      }
      const selectedActivity = activities.find(
        (x: IVariables) => activity === x.activityCode,
      );
      if (!location || !legalForm || !selectedActivity) {
        props.history.push('/business-licence-procedure/error-page');
      } else {
        actions.resultState.update({
          ...formCompanyDetails,
          selectedActivity,
        });
      }
    },
    fromProcessState: {},
    state: {
      mapState: [
        'loggedIn',
        'user',
        'instanceId',
        'businessKey',
        'formCompanyDetails',
        'formSelectActivity',
        'resultState',
        {
          steps: (state: IVariables) => {
            return state.businessLicenseProcedureSteps;
          },
        },
      ],
      mapDispatch: [
        'formRenewEconomicLicence',
        'instanceId',
        'businessKey',
        'resultState',
        'downloadLicenseProcedureDoc',
      ],
    },
  },
];

export default routes;
