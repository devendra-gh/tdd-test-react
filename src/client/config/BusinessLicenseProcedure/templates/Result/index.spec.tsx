import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { MemoryRouter } from 'react-router';
import { IVariables } from '@tamm/app-composer';
import { Viewport, ViewportProvider } from '@tamm/ui-lib-v2-viewport';
import Result from './index';

jest.mock('client/templates/Sidebar', () => () => '');
jest.mock('@tamm/app-composer', () => ({
  withTemplateHooks: (component: any) => component,
}));

// eslint-disable-next-line no-console
console.log = () => {};
console.info = () => {};
console.warn = () => {};
console.error = () => {};

describe('Result', () => {
  let props: IVariables;

  beforeEach(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });

    props = {
      history: {
        push: jest.fn(),
      },
      onChange: jest.fn(),
      onSubmit: jest.fn(),
      i18n: (key: string, object: IVariables) => {
        return object && object.amount ? key + object.amount : key;
      },
      title: 'Result',
      subTitle: 'subTitle',
      description: 'desc',
      locale: 'en',
      steps: [],
      stepStatus: {},
      formSelectActivity: {
        activities: [
          {
            activityCode: 1,
            activity: 'activity',
            activityAr: 'activityAr',
          },
        ],
        activity: 1,
      },
      getInitialState: jest.fn(),
      resultState: {
        selectedActivity: {
          activityCode: 1,
          activity: 'activity',
          activityAr: 'activityAr',
        },
        requirements: [
          {
            RequirementDescEn: 'RequirementDescEn',
            RequirementDescAr: 'RequirementDescAr',
            requirementTypeEn: 'requirementTypeEn',
            AuthorityEn: 'AuthorityEn',
            AuthorityAr: 'AuthorityAr',
          },
        ],
        fees: [],
        loading: true,
        showError: false,
      },
    };
  });

  const renderWithProps = () =>
    render(
      <MemoryRouter>
        <ViewportProvider>
          <Viewport sm md lg xl>
            <Result {...props} />
          </Viewport>
        </ViewportProvider>
      </MemoryRouter>,
    );
  it('should display Loading component when loading is true', async () => {
    const { getByText, container } = renderWithProps();
    expect(container.querySelector('.spinner-wrapper')).toBeInTheDocument();
    expect(getByText(props.subTitle)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display Table component when locale en', async () => {
    props.resultState.showError = false;
    props.resultState.loading = false;
    const { getByText } = renderWithProps();
    expect(getByText('table.heading.licenceSummary')).toBeInTheDocument();
    expect(getByText('activity')).toBeInTheDocument();
  });

  it('should display Table component when locale ar', async () => {
    props.resultState.showError = false;
    props.resultState.loading = false;
    props.locale = 'ar';
    const { getByText } = renderWithProps();
    expect(getByText('table.heading.licenceSummary')).toBeInTheDocument();
    expect(getByText('activityAr')).toBeInTheDocument();
  });

  it('should work when formSelectActivity has bad data', async () => {
    props.resultState.showError = false;
    props.resultState.loading = false;
    props.locale = 'ar';
    props.resultState.selectedActivity = {
      activityCode: 1,
    };
    const { getByText } = renderWithProps();
    expect(getByText('table.heading.licenceSummary')).toBeInTheDocument();
  });

  it('Fee addition must be correct', async () => {
    props.resultState.showError = false;
    props.resultState.loading = false;
    props.locale = 'en';
    props.resultState.fees = [
      { feeAmount: 400, feeDescEn: 'FeeDescEn1', feeDescAr: 'feeDescAr1' },
      { feeAmount: 515, feeDescEn: 'FeeDescEn2', feeDescAr: 'feeDescAr2' },
    ];
    const { getByText } = renderWithProps();
    expect(getByText('global.aed915')).toBeInTheDocument();
  });

  it('Fee addition must be correct ar', async () => {
    props.resultState.showError = false;
    props.resultState.loading = false;
    props.locale = 'ar';
    props.resultState.fees = [
      { feeAmount: 400, feeDescEn: 'FeeDescEn1', feeDescAr: 'feeDescAr1' },
      { feeAmount: 515, feeDescEn: 'FeeDescEn2', feeDescAr: 'feeDescAr2' },
    ];
    const { getByText } = renderWithProps();
    expect(getByText('global.aed915')).toBeInTheDocument();
  });

  it('Clicking back button shound call history.push', async () => {
    const { getByLabelText } = renderWithProps();
    fireEvent.click(getByLabelText('button'));
    expect(props.history.push).toHaveBeenCalled();
  });
});
