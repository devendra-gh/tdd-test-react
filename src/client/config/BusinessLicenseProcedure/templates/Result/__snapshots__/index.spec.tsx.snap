// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result should display Loading component when loading is true 1`] = `
<div
  class="ui-lib-viewport ui-lib-viewport_lg"
>
  <div
    class="container"
  >
    <div
      class="ant-row ui-lib-grid-row ui-lib-grid-row "
      style="margin-left: -10px; margin-right: -10px;"
    >
      <div
        class="ant-col ui-lib-grid-col ui-lib-grid-col-0 ui-lib-grid-col  ui-lib-grid-col-xs-24 ui-lib-grid-col-sm-24 ui-lib-grid-col-md-24 ui-lib-grid-col-lg-16 ui-lib-grid-col-xl-18"
        style="padding-left: 10px; padding-right: 10px;"
      >
        <div
          class=""
        >
          <h3
            class="syb-title"
            style="margin-bottom: 2rem; color: rgb(22, 16, 56);"
          >
            subTitle
          </h3>
          <p
            style="color: rgb(63, 62, 69);"
          >
            desc
          </p>
        </div>
        <div
          style="height: 40px;"
        />
        <div
          class="spinner-wrapper"
        >
          <div
            class="ui-lib-spinner qa-spinner ui-lib-spinner_logo"
          />
        </div>
        <div
          style="height: 40px;"
        />
        <form
          class="ui-lib-form qa-form"
          id=""
        >
          
          <div />
          <div
            class="ui-lib-form__controls"
          >
            <div
              class="ui-lib-form__controls-buttons"
            >
              <button
                aria-label="button"
                class="ui-lib-button qa-button ui-lib-button_secondary ui-lib-button_default ui-lib-button_icon-start ui-lib-button_space-between ui-lib-button_with-icon"
                type="button"
              >
                button.back
                <svg
                  class="ui-lib-button__icon ui-lib-icon"
                  data-key=""
                  height="24"
                  version="1.1"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title />
                  <path
                    d="M22.548,10.561H5.437a.251.251,0,0,1-.165-.438l8.637-7.6a1.44,1.44,0,0,0-1.9-2.162L.828,10.2a2.4,2.4,0,0,0,0,3.6l11.179,9.837a1.44,1.44,0,0,0,1.9-2.161l-8.637-7.6a.251.251,0,0,1,.165-.438H22.548a1.44,1.44,0,0,0,0-2.879Z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </form>
        <div
          style="height: 40px;"
        />
      </div>
      <div
        class="ant-col ui-lib-grid-col ui-lib-grid-col-0 ui-lib-grid-col  ui-lib-grid-col-xs-24 ui-lib-grid-col-sm-24 ui-lib-grid-col-md-24 ui-lib-grid-col-lg-8 ui-lib-grid-col-xl-6"
        style="padding-left: 10px; padding-right: 10px;"
      >
        
      </div>
    </div>
  </div>
</div>
`;
